### ISSUE TITLE
[Feature-request] Allow auxiliary tasks in policies through a unified interface

### ISSUE THREAD

**Issue by PartiallyTyped** [DLR-RM/stable-baselines3#59]
## Motivation
There is a large corpus of work that shows that auxiliary tasks (i.e. anything besides learning a policy and a value function) can help agents get better state representation, e.g. AutoEncoders, forward predictive models, curl etc.

## Proposal
The current implementation of algorithms enables an almost trivial way of incorporating additional tasks through slight modifications of the source code. The proposal is to make it so that changes need to happen in a small segment of the code instead of every function that calls the policy. 

This request is for > 1.2.

## Specifics
Consider for example the ActorCriticPolicy class
The suggestion is to change:
```python
def _get_latent(self, obs: th.Tensor) -> Tuple[th.Tensor, th.Tensor, th.Tensor]:
```
to 
```python
def _get_latent(self, obs: th.Tensor) -> Tuple[th.Tensor, th.Tensor, th.Tensor, Optional[Tuple[th.Tensor,...]]]:
```
and propagate the results where necessary, for example:

```python
    def forward(self, obs: th.Tensor,
                deterministic: bool = False) -> Tuple[th.Tensor, th.Tensor, th.Tensor]:
        
        latent_pi, latent_vf, latent_sde = self._get_latent(obs)

        values = self.value_net(latent_vf)
        distribution = self._get_action_dist_from_latent(latent_pi, latent_sde=latent_sde)
        actions = distribution.get_actions(deterministic=deterministic)
        log_prob = distribution.log_prob(actions)
        return actions, values, log_prob
```

becomes

```python
    def forward(self, obs: th.Tensor,
                deterministic: bool = False) -> Tuple[th.Tensor, th.Tensor, th.Tensor, Optional[Tuple[th.Tensor,...]]]:
        
        latent_pi, latent_vf, latent_sde, other_latent = self._get_latent(obs)

        values = self.value_net(latent_vf)
        distribution = self._get_action_dist_from_latent(latent_pi, latent_sde=latent_sde)
        actions = distribution.get_actions(deterministic=deterministic)
        log_prob = distribution.log_prob(actions)
        return actions, values, log_prob, other_latent
```
and account for this when a rollout is collected.
The consequence here is that when deriving the algorithm, eg. to add an AE, we will only need to derive the `_get_latent` and iirc `train` functions, the former to return the latent of the encoder, the latter to compute the loss of decoder.

**Comment by araffin** [comment]
Related: https://github.com/hill-a/stable-baselines/issues/463

I need to think more about it, but for now I would prefer that users define custom policies and train methods (related to #55 though) rather than changing the default code.

**Comment by PartiallyTyped** [comment]
Perhaps this can be implemented through a low level API as #55 suggests.
